### PR TITLE
Added an optional options parameter for WebSockets

### PR DESCRIPTION
**Pull request by aprock**
This enables overriding origin, and other request headers. Similar to the https://github.com/websockets/ws api.

### CODE REVIEW

**Comment by satya164** [comment]
@aprock Are these part of the spec? What's the use case?

cc @mkonicek

**Comment by satya164** [comment]
Thoughts @ide @mkonicek ?

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 1]
```diff
@@ -465,12 +472,21 @@ - (void)didConnect
   CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Sec-WebSocket-Key"), (__bridge CFStringRef)_secKey);
   CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Sec-WebSocket-Version"), (__bridge CFStringRef)[NSString stringWithFormat:@"%ld", (long)_webSocketVersion]);
 
-  CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Origin"), (__bridge CFStringRef)_url.RCTSR_origin);
-
   if (_requestedProtocols) {
     CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Sec-WebSocket-Protocol"), (__bridge CFStringRef)[_requestedProtocols componentsJoinedByString:@", "]);
   }
 
+  NSString *origin = (NSString *)[_requestedOptions objectForKey: @"origin"];
```
Comment (ide) [inline Libraries/WebSocket/RCTSRWebSocket.m]: Do more graceful type checking here -- `RCTLog` an error and fall back to the default origin if the value is not a string. Also prefer dictionary indexing syntax.

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 2]
```diff
@@ -465,12 +472,21 @@ - (void)didConnect
...
+  if (!origin) {
+    origin = _url.RCTSR_origin;
+  }
+  CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Origin"), (__bridge CFStringRef)origin);
+
+  NSDictionary<NSString *, id> *headers = [_requestedOptions objectForKey: @"headers"];
```
Comment (ide) [inline Libraries/WebSocket/RCTSRWebSocket.m]: Same here with type checking and syntax.

File: Libraries/WebSocket/WebSocketBase.js [hunk 3]
```diff
@@ -33,19 +33,31 @@ class WebSocketBase extends EventTarget {
...
     super();
     this.CONNECTING = 0;
     this.OPEN = 1;
     this.CLOSING = 2;
     this.CLOSED = 3;
 
-    if (!protocols) {
-      protocols = [];
+    if (protocols && !Array.isArray(protocols) && typeof protocols === 'object') {
+      // accept the "options" Object as the 2nd argument
```
Comment (ide) [inline Libraries/WebSocket/WebSocketBase.js]: I don't think we should do this. It complicates the type signature once we try to use more precise types than `any?`. Always specify a value for `protocols` -- much simpler that way.

File: Libraries/WebSocket/WebSocketBase.js [hunk 4]
```diff
@@ -33,19 +33,31 @@ class WebSocketBase extends EventTarget {
   readyState: number;
   url: ?string;
 
-  constructor(url: string, protocols: ?any) {
+  /**
+   * @constructor WebSocket
```
Comment (ide) [inline Libraries/WebSocket/WebSocketBase.js]: Delete this -- this is intrinsically expressed by the code.

File: Libraries/WebSocket/WebSocketBase.js [hunk 5]
```diff
@@ -33,19 +33,31 @@ class WebSocketBase extends EventTarget {
...
+   * @param {String} address Connection address.
```
Comment (ide) [inline Libraries/WebSocket/WebSocketBase.js]: Move the types into the method signature itself.

File: Libraries/WebSocket/WebSocketBase.js [hunk 6]
```diff
@@ -33,19 +33,31 @@ class WebSocketBase extends EventTarget {
...
+   * @param {String|Array} protocols WebSocket protocols.
+   * @param {Object} options Additional connection options.
```
Comment (ide) [inline Libraries/WebSocket/WebSocketBase.js]: Might want to define a typedef for the options so it's clear exactly what options are supported.

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 7]
```diff
@@ -57,7 +60,7 @@ public String getName() {
   }
 
   @ReactMethod
-  public void connect(final String url, final int id) {
+  public void connect(final String url, @Nullable final String protocols, @Nullable final ReadableMap options, final int id) {
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: protocols should be an array, right? Make sure to test this case.

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 8]
```diff
@@ -65,12 +68,46 @@ public void connect(final String url, final int id) {
...
+              // Can be int or double.
+              builder.addHeader(key, String.valueOf(headers.getDouble(key)));
+              break;
+            case String:
+              builder.addHeader(key, headers.getString(key));
+              break;
+            case Map:
+              throw new UnsupportedOperationException("Maps aren't supported header values.");
+            case Array:
+              throw new UnsupportedOperationException("Arrays aren't supported header values.");
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: Can you add a TODO comment to support this? For other APIs, arrays of header values are joined by `", "`.

**Comment by ide** [comment]
This seems useful since there's no notion of origin otherwise.

File: ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java [hunk 9]
```diff
@@ -65,12 +68,46 @@ public void connect(final String url, final int id) {
...
-        .build();
+        .url(url);
 
-    WebSocketCall.create(client, request).enqueue(new WebSocketListener() {
+    if (options != null) {
+      if (options.hasKey("origin")) {
+        builder.addHeader("Origin", options.getString("origin"));
+      }
+
+      if (options.hasKey("headers")) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: The description of this PR talks about overriding origin. The branch above does that. How is this branch related?

Comment (aprock) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: The ws module's options argument separates "origin" from the other header values that can be added to the request. Rather than duplicate their api exactly, this seemed like a reasonable start to allow for request customization.

Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java]: Could you please keep the PR focused on the origin? In my experience, more code always means more code to review, maintain, and fix bugs in.

**Comment by mkonicek** [comment]
Can you please keep the PR focused on the origin and remove the code for handling the headers? Focused PRs are always easier to review and less code means less code to maintain (especially debug when we discover bugs later).

**Comment by aprock** [comment]
@mkonicek customization is now limited to the origin string.

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 10]
```diff
@@ -465,12 +472,16 @@ - (void)didConnect
   CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Sec-WebSocket-Key"), (__bridge CFStringRef)_secKey);
   CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Sec-WebSocket-Version"), (__bridge CFStringRef)[NSString stringWithFormat:@"%ld", (long)_webSocketVersion]);
 
-  CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Origin"), (__bridge CFStringRef)_url.RCTSR_origin);
-
   if (_requestedProtocols) {
     CFHTTPMessageSetHeaderFieldValue(request, CFSTR("Sec-WebSocket-Protocol"), (__bridge CFStringRef)[_requestedProtocols componentsJoinedByString:@", "]);
   }
 
+  NSString *origin = [_requestedOptions objectForKey: @"origin"];
```
Comment (ide) [inline Libraries/WebSocket/RCTSRWebSocket.m]: `_requestedOptions[@"origin"]`

**Comment by ide** [comment]
looking good for the most part

File: Libraries/WebSocket/WebSocketBase.js [hunk 11]
```diff
@@ -33,19 +33,19 @@ class WebSocketBase extends EventTarget {
...
     super();
     this.CONNECTING = 0;
     this.OPEN = 1;
     this.CLOSING = 2;
     this.CLOSED = 3;
 
-    if (!protocols) {
-      protocols = [];
+    if (typeof protocols === 'string') {
+      protocols = [protocols];
```
Comment (satya164) [inline Libraries/WebSocket/WebSocketBase.js]: What if protocols is not an array or undefined?

File: Libraries/WebSocket/WebSocket.js [hunk 12]
```diff
@@ -24,15 +24,16 @@ var CLOSE_NORMAL = 1000;
...
  *
  * See https://developer.mozilla.org/en-US/docs/Web/API/WebSocket
+ * See https://github.com/websockets/ws
  */
 class WebSocket extends WebSocketBase {
   _socketId: number;
   _subs: any;
 
-  connectToSocketImpl(url: string): void {
+  connectToSocketImpl(url: string, protocols: ?any, options: ?any): void {
```
Comment (satya164) [inline Libraries/WebSocket/WebSocket.js]: Why annotate with `?any`? Isn't it better to have `protocols: ?Array<string>, options: ?Object`?

**Comment by satya164** [comment]
Also, what's the use case with the protocols option?

File: Libraries/WebSocket/WebSocketBase.js [hunk 13]
```diff
@@ -33,19 +33,19 @@ class WebSocketBase extends EventTarget {
   readyState: number;
   url: ?string;
 
-  constructor(url: string, protocols: ?any) {
+  constructor(url: string, protocols: ?string | ?Array<string>, options: ?{origin?: string}) {
```
Comment (satya164) [inline Libraries/WebSocket/WebSocketBase.js]: Isn't it better to have protocols in options object? Also, why allow both string and array? Just array should work.

Comment (aprock) [inline Libraries/WebSocket/WebSocketBase.js]: This is according to spec.  The WS spec has an optional protocols argument that can be a string or array of strings.

https://html.spec.whatwg.org/multipage/comms.html#the-websocket-interface

File: Libraries/WebSocket/WebSocketBase.js (hunk 11 above)
Comment (aprock) [inline Libraries/WebSocket/WebSocketBase.js]: added an if for this

**Comment by satya164** [comment]
cc @mkonicek

**Comment by mkonicek** [comment]
The Android changes look OK to me, thanks! @nicklockwood can I merge this? Also, should I cc other people if you're too busy?

**Comment by nicklockwood** [comment]
@mkonicek yep, this looks fine.

@aprock it might be an idea to put up a PR against the SRWebSocket project as well. We've vendored it into React Native, but it will be easier to keep in sync with trunk if this feature is added to the original too.

**Comment by aprock** [comment]
@nicklockwood ah yes, i did see that ancestry. Good thought, I'll PR them as well.

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit
